common/djangoapps/embargo/models.py

```diff
@@ -72,6 +83,256 @@ def embargoed_countries_list(self):
+
+    1) Retrieve the list of whitelisted countries for the course.
+    (If there aren't any, then include every possible country.)
+
+    2) From the initial list, remove all blacklisted countries
+    for the course.
+
+    """
+
+    RULE_TYPE_CHOICES = (
+        (WHITE_LIST, 'Whitelist (allow only these countries)'),
+        (BLACK_LIST, 'Blacklist (block these countries)'),
+    )
+
+    rule_type = models.CharField(
+        max_length=255,
+        choices=RULE_TYPE_CHOICES,
+        default=BLACK_LIST,
+        help_text=ugettext_lazy(
+            u"Whether to include or exclude the given course. "
+            u"If whitelist countries are specified, then ONLY users from whitelisted countries "
+            u"will be able to access the course.  If blacklist countries are specified, then "
+            u"users from blacklisted countries will NOT be able to access the course."
+        )
+    )
+
+    restricted_course = models.ForeignKey(
+        "RestrictedCourse",
+        help_text=ugettext_lazy(u"The course to which this rule applies.")
+    )
+
+    country = models.ForeignKey(
+        "Country",
+        help_text=ugettext_lazy(u"The country to which this rule applies.")
+    )
+
+    @classmethod
+    def cache_key_for_consolidated_countries(cls, course_id):
+        return "{}/allowed/countries".format(course_id)
+
+
+    @classmethod
+    def check_country_access(cls, course_id, country):
+        """
+        Check if the country is either in whitelist or blacklist of countries for the course_id
+
+        Args:
+            course_id (str): course_id to look for
+            country (str): A 2 characters code of country
+            rule_type (str): whitelist or blacklist
+
+        Returns:
+            Boolean
+            True if no country found for the given course and rule type
+            otherwise check given country exists in list
+        """
+
+        blocked = True
+        allowed_countries = cls._get_country_access_list(course_id)
+
+        if country in allowed_countries:
+            blocked = None
+
+        return blocked
+
+
+    @classmethod
+    def _get_country_access_list(cls, course_id):
+        allowed_countries = cache.get(cls.cache_key_for_consolidated_countries(course_id))
+        if not allowed_countries:
+            all_countries = [code[0] for code in list(countries)]
+            black_countries = cls._get_country_access_rules_for_course(course_id, BLACK_LIST)
+            white_countries = cls._get_country_access_rules_for_course(course_id, WHITE_LIST)
+
+            if white_countries and black_countries:
+                allowed_countries = cls._get_country_access_rules_for_course(course_id, WHITE_LIST)
+            elif white_countries:
+                allowed_countries = cls._get_country_access_rules_for_course(course_id, WHITE_LIST)
+            elif black_countries:
+                allowed_countries = list(set(all_countries) - set(black_countries))
```

Comment:
Would this be equivalent?
```
whitelist_countries = set()
blacklist_countries = set()

# Retrieve all rules in one database query, performing the "join" with the Country table
rules_for_course = CountryAccessRule.objects.select_related('country').filter(restricted_course__course_key=course_id)

# Filter the rules into a whitelist and blacklist in one pass
for rule in rules_for_course:
    if rule.rule_type == 'whitelist':
        whitelist_countries.add(rule)
    elif rule.rule_type == 'blacklist':
        blacklist_countries.add(rule)

# If there are no whitelist countries, default to all countries
if not whitelist_countries:
    whitelist_countries = set(code[0] for code in list(countries))

# Consolidate the rules into a single list of countries
# that have access to the course.
return list(whitelist_countries - blacklist_countries)
```

The advantage of doing it this way is that:
1) The uncached case requires one database query instead of two.
2) There are no additional queries to the cache.
3) We don't need to handle each combination of whitelist / blacklist as a special case.